message/validation/validation.go

```diff
@@ -0,0 +1,559 @@
+// Package validation provides functions and structures for validating messages.
+package validation
+
+// validator.go contains main code for validation and most of the rule checks.
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+	specqbft "github.com/bloxapp/ssv-spec/qbft"
+	spectypes "github.com/bloxapp/ssv-spec/types"
+	"github.com/cornelk/hashmap"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"golang.org/x/exp/slices"
+
+	"github.com/bloxapp/ssv/logging/fields"
+	"github.com/bloxapp/ssv/network/commons"
+	"github.com/bloxapp/ssv/networkconfig"
+	"github.com/bloxapp/ssv/operator/duties/dutystore"
+	ssvmessage "github.com/bloxapp/ssv/protocol/v2/message"
+	"github.com/bloxapp/ssv/protocol/v2/ssv/queue"
+	ssvtypes "github.com/bloxapp/ssv/protocol/v2/types"
+	registrystorage "github.com/bloxapp/ssv/registry/storage"
+)
+
+const (
+	// lateMessageMargin is the duration past a message's TTL in which it is still considered valid.
+	lateMessageMargin = time.Second * 3
+
+	// clockErrorTolerance is the maximum amount of clock error we expect to see between nodes.
+	clockErrorTolerance = time.Millisecond * 50
+
+	maxMessageSize             = maxConsensusMsgSize
+	maxConsensusMsgSize        = 8388608
+	maxPartialSignatureMsgSize = 1952
+	allowedRoundsInFuture      = 1
+	allowedRoundsInPast        = 2
+	lateSlotAllowance          = 2
+	signatureSize              = 96
+	maxDutiesPerEpoch          = 2
+)
+
+// ConsensusID uniquely identifies a public key and role pair to keep track of state.
+type ConsensusID struct {
+	PubKey phase0.BLSPubKey
+	Role   spectypes.BeaconRole
+}
+
+// ConsensusState keeps track of the signers for a given public key and role.
+type ConsensusState struct {
+	// TODO: consider evicting old data to avoid excessive memory consumption
+	Signers *hashmap.Map[spectypes.OperatorID, *SignerState]
+}
+
+// GetSignerState retrieves the state for the given signer.
+// Returns nil if the signer is not found.
+func (cs *ConsensusState) GetSignerState(signer spectypes.OperatorID) *SignerState {
+	signerState, ok := cs.Signers.Get(signer)
+	if !ok {
+		return nil
+	}
```

Comment:
this code works, but checking is meaningless since if `!ok` `singerState` will be `nil` anyway.